### DISCUSSION TITLE
ImageSource from a IFeed

### DISCUSSION THREAD

**Discussion by igiona** [unoplatform/uno#15016] [General]
Hi everyone :)

I'm working on a cross-platform application which requires to show a jpeg "stream" coming from a socket.
I started working on the prototype using MVUX and I planned to use my frames as IFeed<ImageSource> such that:

```c#
    public IFeed<ImageSource> CameraFrame => Feed.AsyncEnumerable(CameraFrameSequence);


    private async IAsyncEnumerable<ImageSource> CameraFrameSequence([EnumeratorCancellation] CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            byte[] frame = await GetNextFrame(ct);
            if (frame != null)
            {
                BitmapImage? bmp = null;
                try
                {                    
                    bmp = new BitmapImage();
                    bmp.SetSource(new MemoryStream(frame).AsRandomAccessStream());
                }
                catch (Exception e)
                {
                }

                if (bmp != null)
                {
                    yield return bmp!;
                }
            }
        }
    }
```
On the creation of BitmapImage I get the following Exception:

```
{"The application called an interface that was marshalled for a different thread. (0x8001010E (RPC_E_WRONG_THREAD))"}
This exception was originally thrown at this call stack:
    System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(int) in Marshal.cs
    WinRT.BaseActivationFactory._ActivateInstance<I>() in BaseActivationFactory.cs
    WinRT.ActivationFactory<T>.ActivateInstance<I>() in ActivationFactory.cs
    Microsoft.UI.Xaml.Media.Imaging.BitmapImage.BitmapImage() in BitmapImage.cs
    BroApp.Presentation.MainModel.CameraFrameSequence(System.Threading.CancellationToken) in MainModel.cs
```

I "solved" the problem by delegating the BMP creation to the UI thread:

```c#
bmp = await _dispatcher.ExecuteAsync<BitmapImage>(async token =>
{
    var bmp = new BitmapImage();
    bmp.SetSource(new MemoryStream(frame).AsRandomAccessStream());
    return bmp;
}, ct);
```

It doesn't seem the right thing to do here, and this is reflected also in the poor performance on a mobile device (Android).
The ImageSource is getting there, the camera stream is quite smooth as well but the control is terribly flickering.
I guess because I'm jamming the UI thread...

I'm happy to get any suggestion 😄 
Is this the right approach?
What other way is there to get it to run?
Is there a "uno-way" ?

Thank you for your hard work!
Cheers!
ig

PS: MVUX is impressive!

**Comment by dr1rrb** [comment]
Hey @igiona, thanks for the good feedback :)

Even if I do like the idea to use an `IFeed` as an image stream, I'm wondering if why you are not using a media player with a [mjpeg video stream](https://en.wikipedia.org/wiki/Motion_JPEG)?

**Reply by igiona** [reply]
Thanks your input @dr1rrb , there is no particular reason other that I didn't though about doing in that way, and because I don't know the m-jpeg compression and I'm not familiar with the media-player control.

> From wikipedia
Unlike the video formats specified in [international standards](https://en.wikipedia.org/wiki/International_standard) such as [MPEG-2](https://en.wikipedia.org/wiki/MPEG-2) and the format specified in the JPEG still-picture coding standard, there is no document that defines a single exact format that is universally recognized as a complete specification of “Motion JPEG” for use in all contexts.

Is there any specification for the media-player control, specially about the m-jpeg format it expects?

**Comment by francoistanguay** [comment]
@igiona  I'll let David answer if we have links to info besides https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.mediaplayerelement?view=windows-app-sdk-1.4

For MVUX/IFeed, bear in mind these constructs were created to sow the Loading/Error/Empty/Value states of data as it is being fetched and then shown onto a View. I dont think it would fit your scenario nicely.

**Comment by igiona** [comment]
Thank you @francoistanguay for the heads up and hint.

I tried out (unsuccessfully) somethings on the MediaControlElement.

Here is the latest state:

View
```xaml 
    <MediaPlayerElement
        Width="700"
        MinHeight="700"
        BorderThickness="2"
        Source="{Binding CameraStream}"
        Grid.Row="4" Grid.ColumnSpan="2"
        HorizontalAlignment="Center"
        VerticalAlignment="Center"
        Margin="8"
        AutoPlay="True"
        AreTransportControlsEnabled="True"
      />
```

Model
```c#
    public IState<MediaSource> CameraStream => State<MediaSource>.Value(this, () =>
        {
            //return MediaSource.CreateFromUri(new Uri("https://uno-assets.platform.uno/tests/videos/Mobile_Development_in_VS_Code_with_Uno_Platform_orDotNetMAUI.mp4"));
            return MediaSource.CreateFromStream(_cameraStream.AsRandomAccessStream(), "multipart/x-mixed-replace; boundary=--myboundary");
        }
    );

   MemoryStream _cameraStream = new MemoryStream();
   async Task AddFrame(byte[] frame)
   {
       _logger.Debug(string.Format("New frame {0}", _cameraStream.Length));
       var _cameraStream= _cameraStream;

       if (_cameraStream.Length == 0)
       {
           _cameraStream.Write(Encoding.ASCII.GetBytes("HTTP/1.1 200 OK\r\nContent-Type: multipart/x-mixed-replace; boundary=--myboundary\r\n\r\n"));
       }

       _cameraStream.Write(Encoding.ASCII.GetBytes($"\r\n--myboundary\r\nContent-Type: image/jpeg\r\nContent-Length: {frame.Length}\r\n"));
       _cameraStream.Write(frame);
       //_cameraStream.Write(File.ReadAllBytes(@"C:\test\TEST.jpg"));
       _cameraStream.Write(Encoding.ASCII.GetBytes("\r\n"));

       //await CameraStream.Update(_ => MediaSource.CreateFromStream(stream.AsRandomAccessStream(), "multipart/x-mixed-replace; boundary=--myboundary"), CancellationToken.None);
       await Task.CompletedTask;
   }
```

When using `MediaSource.CreateFromUri` in the CameraStream  initializer, the video plays as expected.
Using the stream doesn't work...
Then I figured out this on your documentation https://platform.uno/docs/articles/controls/MediaPlayerElement.html

```
If you need to set source programmatically (i.e., using _mediaPlayerElement.Source = [source]), please note that only sources created with MediaSource.CreateFromUri() are currently supported
```

@dr1rrb how did you envisioned this to work?

I could create a kind of "local HTTP server" that provides (within the app) the m-jpeg stream from the frames I get, but it seems quite an overhead to me.
Every suggestion is welcomed 😄

**Reply by dr1rrb** [reply]
Hummm yes thinking about the restrictions and how we implemented the `MediaPlayerElement` is makes sense that we support only uri based sources for now :/

The issue is that decompressing a jpeg image is quite a "long" task and I would assume that, unless maybe on high end devices, you will never be able to have a video stream at 25fps using an `Image` control on which you change the `ImageSource` every few ms with a lot of interop involved (From jpeg decoding -which is native- to the re-layouting of the visual tree when you change the source -hopefully cached, but would still implies lot of things-), not even counting the possible stream copy (which is required on wasm).

About starting a server locally, I expect this to work only on skia target. On iOS, Android you will face permission restrictions, and on wasm it's impossible.

For me the only solution then is really to expose a video stream from your server. You mentioned that you get your jpeg from a socket (BTW `TCPClient` is not supported by dotnet on wasm, you have to use HTTP/WebSocket), I would instead encode the video server side and then expose it though an HTTP server.

**Reply by igiona** [reply]
@dr1rrb even if I would provide a MJPEG stream created server side, the client (phone) will have to decompress the jpeg frames anyhow, so I'm not sure why this approach should run smoother?

>About starting a server locally, I expect this to work only on skia target. On iOS, Android you will face permission restrictions, and on wasm it's impossible.

I wouldn't mind the wasm restriction too much, but mobile is a must.
With restrictions, you mean that the OS doesn't allow to listen to an arbitrary TCP/Port ?

>For me the only solution then is really to expose a video stream from your server. You mentioned that you get your jpeg from a socket (BTW TCPClient is not supported by dotnet on wasm, you have to use HTTP/WebSocket), I would instead encode the video server side and then expose it though an HTTP server.

I'm aware of the Socket restriction on Wasm.
Setting up a HTTP server which should stream a local stream (in which format? A HTTP server like Nginx is capable of doing "on-the-fly" encoding?) is a bit too outside of my comfort area of expertise 😄 
But I see the potential on "simply" creating a basic """"HTTP Server"""" for this particular one single task.
The communication protocol for the MJPEG stream doesn't look too complex.
Does the MediaPlayerElement requires the HTTP body to be encoded or compressed, or it can be plain binary?

**Reply by igiona** [reply]
@dr1rrb I've been able to provide a MJPEG stream instead of the bare frames as you suggested above.
On a web browser the HTTP (not s!) stream is displayed correctly, as well as in VLC 🥳 .

The <MediaPlayerElement> though, doesn't play the stream.
On Windows:
- It opens the camera stream socket but after a couple frames it closes it and none of the frames seems to be displayed

On Android
- The player seems to be stuck in loading something, but I don't see any traffic to the socket. Permissions seem to be fine, because when I set the source to "https://uno-assets.platform.uno/tests/videos/Mobile_Development_in_VS_Code_with_Uno_Platform_orDotNetMAUI.mp4" is plays the video.

Is there a way to debug this issue further/deeper ?

**Reply by dr1rrb** [reply]
> so I'm not sure why this approach should run smoother?

Because it will be deferred to the GPU, which will even be able to drop some frames if needed. Also it involves much less processing around that, like for instance once started, the size of the video is known, while when we load a jpeg image we need to re-validate every thing (i.e. confirm the whole visual tree is still valid and doesn't have to be re-draw).

> With restrictions, you mean that the OS doesn't allow to listen to an arbitrary TCP/Port ?

You should be able to listen on arbitrary ports, but I would assume that it requires you to get the "local network" permission, which might be confusing for users why they should give you access to their local network. But it's only hypothesis, it's something that needs to be validated with each store current rules set.

> Does the MediaPlayerElement requires the HTTP body to be encoded or compressed, or it can be plain binary?

Video streams are already compressed (even MJPEG), so there is probably no benefit to re-compress it.

> Setting up a HTTP server which should stream a local stream (in which format? A HTTP server like Nginx is capable of doing "on-the-fly" encoding?) is a bit too outside of my comfort area of expertise 😄
But I see the potential on "simply" creating a basic """"HTTP Server"""" for this particular one single task.

At some point I think that it would be easier to setup a streaming server than trying to encode the stream yourself. And the fact that your stream is stopping after few frames is one of the reason why 😜. (And it will be really hard to debug since it's the GPU that is decoding the stream). If I had to do that, I would try to use VLC as streaming server: https://www.videolan.org/streaming-features.html. But I didn't tried since a while, might not be the best solution for you case (I don't known what's the source you are trying to stream).

BTW : I suggested the mjpeg protocol as it was the closest to what you did, but it's not the most efficient. If you have opportunity to select the protocol, you should use something like H.264.

**Reply by igiona** [reply]
The device which streams the camera frames has limited processing power (and no GPU), and often doesn't have internet connection, but simply uses a local network to communicate.
I'm afraid that it won't be able to sustain a heavier encoding but it has to evaluated, but it looks like we have enough BW for the MJPEG stream, and in any case the fps isn't too much relevant either.

Assuming that a MJPEG protocol would be the way to go... from here https://platform.uno/docs/articles/controls/MediaPlayerElement.html it doesn't seem to be supported?! 🤔 

I played around with a ffmpeg and tried to setup a MJPEG stream from a file, and the player doesn't play that stream either:
`ffmpeg -re -i /c/tmp/file.mjpeg -listen 1 -f mpjpeg -movflags frag_keyframe http://localhost:1234`
and
`ffmpeg -re -i /c/tmp/file.mjpeg -listen 1 -f mpjpeg http://localhost:1234`
 
but it does play the mp4 stream:
`ffmpeg -re -i /c/tmp/file.mjpeg -listen 1 -f mp4 -movflags frag_keyframe http://localhost:1234`

Given that my MJPEG stream can be played by the browser, is there a chance to embed a browser view into the app to play the video stream, if the player really doesn't support this?

**Comment by zipswich** [comment]
Not sure how relevant this is, but here is how I did for a published app powered by Uno. I extracted the relevant code from multiple locations:
The page layout has the Imge control:
```
                <Image 
                    x:Name="img" 
                    HorizontalAlignment="Center" 
                    VerticalAlignment="Center" 
                    Stretch="UniformToFill"/>
```

The code behind:

```
async void onMJPEGDeocderNewFrame(byte[] abJPEG)
{
        await img.Dispatcher.RunAsync(CoreDispatcherPriority.Normal,
            async () =>
        {
            BitmapImage biNew = new BitmapImage();
            using (Stream stream = new MemoryStream(abJPEG))
            {
                await biNew.SetSourceAsync(stream);
            }
            if (biNew == null || biNew.PixelWidth == 0 || biNew.PixelHeight == 0)
            {
               //Do something to log the error
            }
            else
            {
                img.Source = biNew;
                ++iFrameCount;
            }
        });
}
```

**Reply by igiona** [reply]
Thank you for your suggestion @zipswich !

I actually first went for a similar solution, but instead of setting `img.source` I used a binding on the Source of the image.
Also in this way I experienced the same flickering on the image refresh I afterwards encountered using the IFeed solution above.

I tried your approach on my app, but with the Uno-5 + net8+&MVUX default VS Solution I cannot access the UI element as you did `img.Source`  from the "ViewModel".
Maybe are you using MVVM?

On the `Microsoft.UI.Xaml.Media.Imaging.Bitmap` object the method `SetSourceAsync` seems now to require a random access stream, hence I would have to:
```c#
using (Stream stream = new MemoryStream(abJPEG))
{
    await bmp.SetSourceAsync(stream.AsRandomAccessStream());
}
```

Additionally, the IDispatcher dispatcher I get in the Model doesn't provide a `RunAsync` mehod with the possibility to specifiy a priority.
It seems that it provides only a `ExecuteAsync` method instead.

**Comment by zipswich** [comment]
@igiona I am using Uno UWP, not WinUI. I explored WinUI and decided to hold off for now. 

Here is my code in its entirety
```
#if NETFX_CORE
                using (IRandomAccessStream stream = new MemoryStream(abJPEG).AsRandomAccessStream())
                {
                    await biNew.SetSourceAsync(stream);
                }
#else
                //Only heads using Uno can use the following method.
                using (Stream stream = new MemoryStream(abJPEG))
                {
                    await biNew.SetSourceAsync(stream);
                }
#endif
```
I have a UserControl for streaming shared by many pages. 
I occasionally make view models. I usually bind to the page class (in this case, a UserControl class) directly. This approach has worked out very well for me.

Here is an example:
Xaml: 
```
    <Button
        x:Name="btTalk" 
        IsEnabled="{x:Bind bIsProVer}"
        ToolTipService.ToolTip="{x:Bind ConverterParameter=hold_down_to_talk, Converter={StaticResource XRConverter}}"
        Visibility="{x:Bind bTalkable, Converter={StaticResource BooleanToVisibilityConverter}}">
        <SymbolIcon Symbol="Microphone"/>
    </Button>
```
code behind in `public sealed partial class UserControlDeviceView : UserControl, INotifyPropertyChanged`:
```
        public bool bIsProVer
        {
            get
            {
                return Utility.IsAppProVersion(App.iTrialDuration);
            }
        }


        bool _bTalkable = false;
        public bool bTalkable
        {
            get { return _bTalkable; }
            set
            {
                if (_bTalkable != value)
                {
                    _bTalkable = value;
                    NotifyPropertyChanged();
                }
            }
        }
```

I am not sure if this has much relevance to your case.

**Comment by igiona** [comment]
Thanks to the inputs of @jeromelaban , I have been able to solve this issue.
Instead of refreshing the Image.Source at every frame, causing the UI flickering, I set it only once via `WritableBitmap`.
Then I decode the frames via `SkiaSharp` and I write the decoded pixel stream to the `WritableBitmapin` in the UI thread at every new JPEG frame.
The result is a very smooth video stream on all platform I tested: Android, iOS and Windows.

Code snippet:

```c#
    private WriteableBitmap? _bitmap;
    private SKImageInfo? _cameraFrameInfo;

    public IState<ImageSource> CameraFrame => State<ImageSource>.Empty(this);

    private async Task OnCameraFrameReceived(byte[] frame)
    {
        if (_bitmap == null || _cameraFrameInfo == null)
        {
            // Decode the first frame to gather the size information and update the camera-frame image
            var jpeg = SKBitmap.Decode(frame);
            _cameraFrameInfo = new SKImageInfo(jpeg.Width, jpeg.Height, SKColorType.Bgra8888, SKAlphaType.Opaque);
            await _dispatcher.ExecuteAsync(() => _bitmap = new(jpeg.Width, jpeg.Height));
            await CameraFrame.Update(_ => _bitmap, CancellationToken.None);
        }

        _bitmap?.DispatcherQueue?.TryEnqueue(Microsoft.UI.Dispatching.DispatcherQueuePriority.Normal, () => // TODO Verify if on some devices DispatcherQueuePriority.High is required
        {
            var decodedJpeg = SKBitmap.Decode(frame, _cameraFrameInfo.Value);
            using (var destinationPixels = _bitmap.PixelBuffer.AsStream())
            {
                destinationPixels.Write(decodedJpeg.Bytes);
            }
            _bitmap.Invalidate();
        });
    }
```
